ui/apps/dashboard/src/components/Billing/Plans/UpgradeButton.tsx

```diff
@@ -0,0 +1,125 @@
+'use client';
+
+import { useState } from 'react';
+import { useRouter } from 'next/navigation';
+import { NewButton } from '@inngest/components/Button';
+import { toast } from 'sonner';
+
+import CheckoutModal, { type CheckoutItem } from '@/components/Billing/Plans/CheckoutModal';
+import ConfirmPlanChangeModal from '@/components/Billing/Plans/ConfirmPlanChangeModal';
+import { pathCreator } from '@/utils/urls';
+import { PlanNames, isEnterprisePlan, type Plan } from './utils';
+
+type ChangePlanArgs = {
+  item: CheckoutItem;
+  action: 'upgrade' | 'downgrade' | 'cancel';
+};
+
+export default function UpgradeButton({
+  plan,
+  currentPlan,
+  onPlanChange,
+}: {
+  plan: Plan;
+  currentPlan: Plan;
+  onPlanChange: () => void;
+}) {
+  const router = useRouter();
+  const [checkoutData, setCheckoutData] = useState<{
+    action: 'upgrade' | 'downgrade' | 'cancel';
+    items: CheckoutItem[];
+  }>();
+  const showCheckoutModal = checkoutData?.action === 'upgrade';
+  const showChangePlanModal =
+    checkoutData?.action === 'downgrade' || checkoutData?.action === 'cancel';
+
+  const currentPlanName = currentPlan.name;
+  const cardPlanName = plan.name;
+  const currentPlanAmount = currentPlan.amount;
+  const cardPlanAmount = plan.amount;
+
+  const isEnterprise = isEnterprisePlan(currentPlan);
+  const isEnterpriseCard = cardPlanName === PlanNames.Enterprise;
+  const isFreeCard = cardPlanName === PlanNames.Free;
+
+  const isLegacyPlan = !isEnterprisePlan(currentPlan) && !(currentPlan.name in PlanNames);
+
+  const isActive =
+    currentPlanName === cardPlanName || (cardPlanName === PlanNames.Enterprise && isEnterprise);
+
+  const isLowerPlan = (() => {
+    if (isEnterprise) {
+      // If the current plan is Enterprise, all non-Enterprise plans are lower
+      return !isEnterprisePlan(plan);
+    }
+    // For legacy plans, only Free Tier is a downgrade
+    if (isLegacyPlan) {
+      return isFreeCard;
+    }
+    // For non-enterprise plans, compare the amounts
+    return cardPlanAmount < currentPlanAmount;
+  })();
+
+  const buttonLabel = isActive
+    ? 'My Plan'
+    : isEnterpriseCard
+    ? 'Get in touch'
+    : isLowerPlan
+    ? 'Downgrade'
+    : 'Upgrade';
+
+  const onClickChangePlan = ({ item: { planID, name, amount }, action }: ChangePlanArgs) => {
+    setCheckoutData({ items: [{ planID, name, quantity: 1, amount }], action });
+  };
+
+  const onChangePlanSuccess = () => {
+    setCheckoutData(undefined);
+    onPlanChange();
+    router.refresh();
+    toast.success(`Plan changed successfully`);
+  };
+
+  return (
+    <div className="my-8">
+      <NewButton
+        className="w-full"
+        label={buttonLabel}
+        disabled={isActive}
```

Comment:
Nit: It's kinda weird that there's a permanently disabled button on the active plan. But that's in the design so it's a design convo

Reply:
I personally don't find it weird, this is way better than the current design. This is a common pattern